src/wp-admin/includes/update-core.php

```diff
@@ -1708,3 +1750,123 @@ function _upgrade_590_force_deactivate_incompatible_plugins() {
 	}
 }
+
+/**
+ * The upgrader uses the classes in the outgoing version of WordPress.
+ *
+ * If a class is renamed in the incoming version, the upgrader will no longer
+ * be able to find the outgoing class when it is needed.
+ *
+ * This will cause a fatal error. To prevent this, we can preload the
+ * outgoing classes that we need for the upgrade into memory.
+ *
+ * This MUST be done at the beginning of `update_core()`.
+ *
+ * @since 6.0.0
+ *
+ * @global WP_Filesystem_Base $wp_filesystem  WordPress filesystem subclass.
+ * @global string             $wp_version     The current WordPress version.
+ *
+ * @param string $wp_base                     Base path of the WordPress installation.
+ * @param array  $preload_paths               An array of paths to preload, relative to $wp_base.
+ * @param array  $skip_paths                  An array of paths to skip when preloading, relative to $wp_base.
+ * @return true|WP_Error True on success, or WP_Error on failure.
+ */
+function preload_outgoing_version_files( $wp_base, array $preload_paths, $skip_paths = array() ) {
+	global $wp_filesystem, $wp_version;
+
+	if ( ! $wp_filesystem || ! is_object( $wp_filesystem ) ) {
+		return new WP_Error( 'fs_unavailable', __( 'Could not access filesystem.' ) );
+	}
+
+	if ( empty( $preload_paths ) ) {
+		return new WP_Error( 'empty_paths', __( 'The preload paths list is empty.' ) );
+	}
+
+	$errors = new WP_Error();
+
+	// Preload files.
+	$dangerous_paths = array( '', '.', '/' );
+
+	foreach ( $preload_paths as $path ) {
+		if ( ! is_string( $path ) ) {
+			$errors->add(
+				'invalid_path',
+				/* translators: %s: The invalid path. */
+				sprintf( __( '%s is not a valid path.' ), print_r( $path, true ) )
+			);
+			continue;
+		}
+
+		if ( in_array( trim( $path ), $dangerous_paths, true ) ) {
+			$errors->add(
+				'dangerous_path',
+				/* translators: %s: The path to the file or directory. */
+				sprintf( __( '%s is a dangerous path.' ), $path )
+			);
+			continue;
+		}
+
+		$fullpath = trailingslashit( $wp_base ) . $path;
+
+		if ( ! $wp_filesystem->exists( $fullpath ) ) {
+			$errors->add(
+				'path_does_not_exist',
+				/* translators: %s: The path to the file or directory. */
+				sprintf( __( '%s does not exist.' ), $fullpath )
+			);
+			continue;
+		}
+
+		// Skip paths.
+		if ( ! empty( $skip_paths ) ) {
+			foreach ( $skip_paths as $skip_path ) {
+				if ( $path === $skip_path[0] ) {
+					// No version check is needed. Skip immediately.
+					if ( ! isset( $skip_path[1], $skip_path[2] ) ) {
+						continue 2;
+					}
+
+					// Run the version check. Skip on true.
+					if ( version_compare( $wp_version, $skip_path[1], $skip_path[2] ) ) {
+						continue 2;
+					}
+				}
+			}
+		}
+
+		// This is a directory. Recurse.
+		if ( $wp_filesystem->is_dir( $fullpath ) ) {
+			// Get the file or subdirectory names.
+			$contents = array_keys( $wp_filesystem->dirlist( $fullpath, false, false ) );
```

Comment:
I'll split this out to first check if `dirlist()` has returned `false`. If it has, I can add a WP_Error for `dirlist_not_returned`, then `continue`.